docs/standard/native-interop/tutorial-custom-marshaller.md

```diff
@@ -0,0 +1,278 @@
+
+```csharp
+namespace CustomMarshalling
+{
+    [CustomMarshaller(typeof(string), MarshalMode.Default, typeof(Utf32StringMarshaller))]
+    [CustomMarshaller(typeof(string), MarshalMode.ManagedToUnmanagedIn, typeof(ManagedToUnmanagedIn))]
+    internal static unsafe class Utf32StringMarshaller
+    {
+        //
+        // Stateless functions removed
+        //
+
+        public ref struct ManagedToUnmanagedIn
+        {
+            public static int BufferSize => 0x100;
+
+            private uint* _unmanagedValue;
+            private bool _allocated; // Used stack alloc or allocated other memory
+
+            public void FromManaged(string? managed, Span<byte> buffer)
+                => throw new NotImplementedException();
+
+            public uint* ToUnmanaged()
+                => throw new NotImplementedException();
+
+            public void Free()
+                => throw new NotImplementedException();
+        }
+    }
+}
+```
+
+The first of our two native functions can now be called using our UTF-32 string marshallers. The below declaration uses the `LibraryImport` attribute, just like `DllImport`, but relies upon the `MarshalUsing` to instruct the source generator what marshaller to use when calling the native function. There is no need to clarify if the stateless or stateful marshaller should be used. This is handled by the implementer defining the "mode" on the marshaller. The source generator will select the most appropriate marshaller based on the context in which the `MarshalUsing` is applied, `MarshalMode.Default` being the fallback.
+
+```csharp
+// extern "C" DLL_EXPORT void STDMETHODCALLTYPE PrintString(char32_t* chars);
+[LibraryImport(LibName)]
+internal static partial void PrintString([MarshalUsing(typeof(Utf32StringMarshaller))] string s);
+```
+
+## Customize marshalling for a user-defined type
+
+Marshalling a user-defined type requires defining not only the marshalling logic, but also the type in C# to marshal to/from. Let's recall the native type we are trying to marshal and define what it would ideally look like in C#.
+
+```cpp
+struct error_data
+{
+    int code;
+    bool is_fatal_error;
+    char32_t* message;    /* UTF-32 encoded string */
+};
+```
+
+An `int` is the same size in both modern C++ and in .NET. A `bool` is the canonical example for a boolean value in .NET. Building on top of `Utf32StringMarshaller`, we can marshal `char32_t*` as a .NET `string`. Accounting for .NET style the result is the following definition in C#.
+
+```csharp
+struct ErrorData
+{
+    public int Code;
+    public bool IsFatalError;
+    public string? Message;
+}
+```
+
+Following the naming pattern, we will name the marshaller `ErrorDataMarshaller`. Instead of creating a default marshaller, we will only define marshallers for some modes. This will demonstrate the case where if a mode isn't provided, the source generator will fail. Let's start with the "in" direction.
```

Comment:
```suggestion
Following the naming pattern, name the marshaller `ErrorDataMarshaller`. Instead of creating a default marshaller, you'll only define marshallers for some modes. In this case, if a mode isn't provided, the source generator will fail. Start with defining a marshaller for the "in" direction.
```